### PR TITLE
Optimize updates via physical object import

### PR DESCRIPTION
**Pull request by djjuhasz**
Improve preformance when updating physical object data via csv importer

- Add --debug (-d) flag to importer and CLI task
- Output timing data for various subroutines when debug flag set
- Add quickUpdate() method to QubitPhysicalObject model
- Improve unit tests and code testability

### CODE REVIEW

File: lib/PhysicalobjectCsvImporter.class.php [hunk 1]
```diff
@@ -421,56 +441,191 @@ public function savePhysicalobjects($data)
...
+
+    return $output;
+  }
+
+  public function reportTimes()
+  {
+    if (!$this->getOption('debug'))
+    {
+      return sprintf('Total import time: %01.2fs',
+        $this->getElapsedTime('total')).PHP_EOL;
```
Comment (mcantelon) [inline lib/PhysicalobjectCsvImporter.class.php]: Nitpick: I've been putting spaces around the `.` concat operator... not sure if we have a standard for that or not though.

File: lib/model/QubitPhysicalObject.php [hunk 2]
```diff
@@ -326,4 +326,86 @@ public function getRelationsAsArray(int $typeId = null)
...
+    {
+      $connection = QubitTransactionFilter::getConnection();
+    }
+
+    foreach ($updateCols as $name)
+    {
+      if (!empty($data[$name]) && $data[$name] != $this->$name)
+      {
+        $doUpdate = true;
+        $newvals[$name] = $data[$name];
```
Comment (mcantelon) [inline lib/model/QubitPhysicalObject.php:365]: Can't see where this array gets initialized...

File: lib/model/QubitPhysicalObject.php [hunk 3]
```diff
@@ -326,4 +326,86 @@ public function getRelationsAsArray(int $typeId = null)
 
     return $results->fetchAll(PDO::FETCH_ASSOC);
   }
+
+  /**
+   * An quick update of an existing physical_object database row
+   *
+   * If passed data wouldn't alter the current db data, return false without
```
Comment (mcantelon) [inline lib/model/QubitPhysicalObject.php:334]: Nice!

File: lib/model/QubitPhysicalObject.php [hunk 4]
```diff
@@ -326,4 +326,86 @@ public function getRelationsAsArray(int $typeId = null)
...
+        $doUpdate = true;
+        $newvals[$name] = $data[$name];
+      }
+      else
+      {
+        $newvals[$name] = $this->$name;
+      }
+    }
+
+    if (!$doUpdate)
```
Comment (mcantelon) [inline lib/model/QubitPhysicalObject.php:373]: Does this indicate there was an error updating? If so maybe should be an exception? Dunno...

File: test/phpunit/PhysicalobjectCsvImporterTest.php [hunk 5]
```diff
@@ -423,102 +425,29 @@ public function testDoImportNoFilenameException()
...
     $this->assertSame(explode(',', $this->csvHeader), $importer->getHeader());
-    $this->assertSame($this->getCsvRowAsAssocArray(1), $importer->getRow(1));
-    $this->assertSame(2, $importer->countRowsImported());
-    $this->assertSame(4, $importer->countRowsTotal());
+    $this->assertSame($this->getCsvRowAsAssocArray(1), $records->fetchOne());
+    $this->assertSame(3, $importer->countRowsTotal());
   }
 
-  public function testDoImportWithSetHeader()
+  public function testLoadCsvDatWithSetHeader()
```
Comment (mcantelon) [inline test/phpunit/PhysicalobjectCsvImporterTest.php]: Nitpick: Is "Dat" short for data or is it something else?

File: test/phpunit/PhysicalobjectCsvImporterTest.php [hunk 6]
```diff
@@ -645,6 +574,67 @@ public function testMatchExistingRecordsThrowsExceptionOnMultiMatch()
...
+      Save physical object: 0.00s
+      Save keymap:          0.00s
+      Update IO relations:  0.00s
+  Progress reporting:       0.00s
+---------------------------------
+Total import time:          0.00s
+
+EOM;
+
+    $this->assertSame($expectedOutput, $importer->reportTimes());
```
Comment (mcantelon) [inline test/phpunit/PhysicalobjectCsvImporterTest.php:610]: Slick!

**Review by mcantelon — APPROVED** [review]
A few nitpicks/questions, but otherwise looks good!

File: lib/PhysicalobjectCsvImporter.class.php (hunk 1 above)
Comment (djjuhasz) [reply lib/PhysicalobjectCsvImporter.class.php]: [PSR-2-R](https://github.com/php-fig-rectified/fig-rectified-standards/blob/master/PSR-2-R-coding-style-guide-additions.md) recommends a space before and after the concat operator, so I agree we should use that convention.

File: test/phpunit/PhysicalobjectCsvImporterTest.php (hunk 5 above)
Comment (djjuhasz) [reply test/phpunit/PhysicalobjectCsvImporterTest.php]: Yeah, that's a typo. 😊  It should be "Data". :)

File: lib/model/QubitPhysicalObject.php (hunk 2 above)
Comment (djjuhasz) [reply lib/model/QubitPhysicalObject.php:365]: Good catch, I'll add a line to initialize `$newvals` it at the top of the method.

File: lib/model/QubitPhysicalObject.php (hunk 4 above)
Comment (djjuhasz) [reply lib/model/QubitPhysicalObject.php:373]: No, not an error. If `$doUpdate` is false the incoming data matches the data already in the database, so there's no need to do an update.  I'll add a comment to the code to make this clearer.

**Comment by djjuhasz** [comment]
Merged to qa/2.6.x in 8d6ca45b249d5c929f66453b43ad23c60c759aff

**Pull request closed without merging**